### PR TITLE
feat: progress? to say which theorem has been used

### PR DESCRIPTION
**Pull request by RaitoBezarius**
Like `simp?` and similar, `progress?` will provide what theorem was used to make the goal progress.

TODO: 

- [x] grab original stx elements
- [x] show `progress ...` instead of `...` during the print.
- [x] make code action work on this fragment

### CODE REVIEW

File: backends/lean/Base/Progress/Progress.lean [hunk 1]
```diff
@@ -370,14 +373,22 @@ def evalProgress (args : TSyntax `Progress.progressArgs) : TacticM Unit := do
       Arith.scalarTac false
     else
       throwError "Not a linear arithmetic goal"
-  progressAsmsOrLookupTheorem keep withArg ids splitPost (
+  let usedTheorem ← progressAsmsOrLookupTheorem keep withArg ids splitPost (
     withMainContext do
     trace[Progress] "trying to solve assumption: {← getMainGoal}"
     firstTac [assumptionTac, scalarTac])
-  trace[Diverge] "Progress done"
```
Comment (RaitoBezarius) [inline backends/lean/Base/Progress/Progress.lean:377]: @sonmarcho here, I think it was meant to be Progress, I renamed it even though it has nothing to do with this PR, is this fine or should I split the change?

Comment (sonmarcho) [reply backends/lean/Base/Progress/Progress.lean:377]: Yes, that's fine, thanks! :)

File: backends/lean/Base/Progress/Progress.lean [hunk 2]
```diff
@@ -370,14 +376,26 @@ def evalProgress (args : TSyntax `Progress.progressArgs) : TacticM Unit := do
...
+  trace[Progress] "Progress done"
+  return {
+    usedTheorem := usedTheorem
```
Comment (sonmarcho) [inline backends/lean/Base/Progress/Progress.lean:385]:
```suggestion
    usedTheorem
```

Comment (sonmarcho) [reply backends/lean/Base/Progress/Progress.lean:385]: This is correct Lean code right?

**Review by sonmarcho — APPROVED** [review]
This looks good! I have a minor suggestion, but this put aside you're good to merge :)

**Comment by sonmarcho** [comment]
However it seems that it reordered some definitions, leading to some diff in the files generated by Aeneas: this is slightly suspicious.

File: backends/lean/Base/Progress/Progress.lean (hunk 2 above)
Comment (RaitoBezarius) [reply backends/lean/Base/Progress/Progress.lean]: Done, working!

**Comment by RaitoBezarius** [comment]
@sonmarcho feel free to merge :)

**Pull request merged**